### PR TITLE
[gsi_web] Do not initialize CodeClient if scopes are empty.

### PR DESCRIPTION
**Pull request by ditman**
When initializing a `CodeClient`, it is mandatory to pass at least one `scope` (or the JS initialization will crash).

This PR ensures that the `CodeClient` is not created unless `initialScopes.isNotEmpty`, and lets the user know when attempting to use said client that it hasn't been initialized properly.

## Issues

* Fixes: https://github.com/flutter/flutter/issues/139382
* Fixes: https://github.com/flutter/flutter/issues/62474
* Closes: https://github.com/flutter/cocoon/pull/3304 (makes this PR unneeded)

## Tests

* Tested locally with `cocoon`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/google_sign_in/google_sign_in_web/CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,8 @@
-## 0.12.2
+## 0.12.3
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in_web/CHANGELOG.md:1]: `0.12.2+1`?

**Review by stuartmorgan — COMMENTED** [review]
I haven't paged the whole flow back in here, but what if someone creates with no scopes, but then adds scopes later. Shouldn't we create the client then? Or does that already happen?

**Comment by ditman** [comment]
> I haven't paged the whole flow back in here, but what if someone creates with no scopes, but then adds scopes later. Shouldn't we create the client then? Or does that already happen?

This seems doable with the "TokenClient"; there's an ["OverridableTokenClientConfig" object](https://developers.google.com/identity/oauth2/web/reference/js-reference#OverridableTokenClientConfig) that lets you redefine scopes later, but for the CodeClient flow there's not such a thing, the [`requestCode` method of the CodeClient](https://developers.google.com/identity/oauth2/web/reference/js-reference#CodeClient) doesn't support options.

(However, there's a new `include_granted_scopes` property in the CodeClient initialization)

I suspect the way the CodeClient is intended to be used is to re-create the client with `include_granted_scopes` in order to add more scopes to what has already been authorized, but that doesn't jive well with our "init" method.

Maybe the plugin should re-create the code-client as needed with options right before calling `getCode`, but for now, users will have to request all `scopes` beforehand.

File: packages/google_sign_in/google_sign_in_web/lib/src/gis_client.dart [hunk 2]
```diff
@@ -291,7 +293,12 @@ class GisSdkClient {
   /// Requests a server auth code per:
   /// https://developers.google.com/identity/oauth2/web/guides/use-code-model#initialize_a_code_client
   Future<String?> requestServerAuthCode() async {
-    _codeClient.requestCode();
+    assert(_codeClient != null,
+        'CodeClient not initialized correctly. Ensure the `scopes` list passed to `init()` or `initWithParams()` is not empty!');
```
Comment (stuartmorgan) [inline packages/google_sign_in/google_sign_in_web/lib/src/gis_client.dart:297]: Can we have a TODO with an issue link here to track re-creating `_codeClient` when adding scopes?

**Review by stuartmorgan — APPROVED** [review]
LGTM with TODO nit.

**Comment by ditman** [comment]
@stuartmorgan if all the tests pass locally I'm going to land this with the tree on red. I don't think it's going to fix the tree, but... it might kick a lucky build and fix. That way we can continue working on google_sign_in_web without worrying about publishing the bad version (the migration to `package:web`, for example).

**Pull request merged**